File: python/src/application.py
import logging

from analyzer import Gemini
from config import settings
from models import Decision, Result
from storage import Checkpoint, CSVParser, CSVWriter, JSONParser

logger = logging.getLogger(__name__)


class Application:
    def __init__(self):
        self._analyzer = None

    @staticmethod
    def _build_error_result(e: Exception, context: str = "") -> Result:
        error_msg = str(e)

        if isinstance(e, FileNotFoundError):
            logger.error(f"Required file not found{f' during {context}' if context else ''}: {e}")
            return Result(success=False, error_type="file_not_found", error_message=error_msg)

        if isinstance(e, ValueError):
            logger.error(f"Invalid data format{f' during {context}' if context else ''}: {e}")
            return Result(success=False, error_type="invalid_format", error_message=error_msg)

        if isinstance(e, PermissionError):
            logger.error(f"Permission denied{f' during {context}' if context else ''}: {e}")
            return Result(success=False, error_type="permission_denied", error_message=error_msg)

        logger.error(
            f"Unexpected error{f' during {context}' if context else ''}: {e}", exc_info=True
        )
        return Result(
            success=False,
            error_type="unexpected_error",
            error_message="An unexpected error occurred.",
        )

    @property
    def analyzer(self) -> Gemini:
        if self._analyzer is None:
            self._analyzer = Gemini()
            logger.info("Gemini analyzer initialized")
        return self._analyzer

    def extract_tweets(self) -> Result:
        try:
            logger.info(f"Reading tweets from {settings.tweets_archive_path}")
            parser = JSONParser(settings.tweets_archive_path)
            tweets = parser.parse()

            logger.info(f"Extracted {len(tweets)} tweets, writing to CSV")
            with CSVWriter(settings.transformed_tweets_path) as writer:
                writer.write_tweets(tweets)

            logger.info(
                f"Successfully wrote {len(tweets)} tweets to {settings.transformed_tweets_path}"
            )
            return Result(success=True, count=len(tweets))
        except Exception as e:
            return self._build_error_result(e, context="extraction")

    def analyze_tweets(self) -> Result:
        try:
            logger.info(f"Loading tweets from {settings.transformed_tweets_path}")
            parser = CSVParser(settings.transformed_tweets_path)
            tweets = parser.parse()

            if not tweets:
                logger.warning("No tweets found to analyze")
                return Result(success=True, count=0)

            logger.info(f"Loaded {len(tweets)} tweets for analysis")

            analyzed_count = 0
            with Checkpoint(settings.checkpoint_path) as checkpoint:
                start_index = checkpoint.load()
                logger.info(f"Resuming from tweet index {start_index}")

                with CSVWriter(settings.processed_results_path, append=True) as writer:
                    for i in range(start_index, len(tweets), settings.batch_size):
                        batch = tweets[i : i + settings.batch_size]
                        batch_num = (i // settings.batch_size) + 1
                        total_batches = (
                            len(tweets) + settings.batch_size - 1
                        ) // settings.batch_size

                        logger.info(
                            f"Processing batch {batch_num}/{total_batches} "
                            f"(tweets {i+1}-{min(i+len(batch), len(tweets))} of {len(tweets)})"
                        )

                        for tweet in batch:
                            try:
                                result = self.analyzer.analyze(tweet)
                                logger.debug(f"Tweet {tweet.id}: {result.decision.value}")
                                analyzed_count += 1

                                if result.decision == Decision.DELETE:
                                    writer.write_result(result)
                            except Exception as e:
                                logger.error(
                                    f"Failed to analyze tweet {tweet.id}: {e}", exc_info=True
                                )
                                return Result(
                                    success=False,
                                    count=analyzed_count,
                                    error_type="analysis_failed",
                                    error_message=str(e),
                                )

                        checkpoint.save(i + len(batch))
                        logger.info(f"Checkpoint saved at index {i + len(batch)}")

            logger.info(f"Analysis complete. Results written to {settings.processed_results_path}")
            return Result(success=True, count=analyzed_count)
        except Exception as e:
            return self._build_error_result(e, context="analysis")

File: java/src/main/java/com/benx421/tweetaudit/application/Application.java
package com.benx421.tweetaudit.application;

import java.io.IOException;
import java.util.List;
import java.util.logging.Logger;

import com.benx421.tweetaudit.analyzer.GeminiAnalyzer;
import com.benx421.tweetaudit.analyzer.TweetAnalyzer;
import com.benx421.tweetaudit.config.Settings;
import com.benx421.tweetaudit.models.AnalysisResult;
import com.benx421.tweetaudit.models.Decision;
import com.benx421.tweetaudit.models.Tweet;
import com.benx421.tweetaudit.storage.CSVWriter;
import com.benx421.tweetaudit.storage.Checkpoint;
import com.benx421.tweetaudit.storage.ParserType;
import com.benx421.tweetaudit.storage.TweetParser;

/**
 * Orchestrates tweet extraction and analysis workflows.
 * Coordinates the parser, analyzer, writer, and checkpoint components.
 */
public final class Application {

  private static final Logger LOGGER = Logger.getLogger(Application.class.getName());

  private final TweetAnalyzer analyzer;
  private final Settings settings;
  private final Checkpoint checkpoint;

  public Application(Settings settings) {
    if (settings == null) {
      throw new IllegalArgumentException("Settings cannot be null");
    }
    this.settings = settings;
    this.analyzer = GeminiAnalyzer.fromSettings(settings);
    this.checkpoint = new Checkpoint(settings.checkpointPath());
  }

  // Package-private constructor for testing with mock analyzer
  Application(Settings settings, TweetAnalyzer analyzer) {
    if (settings == null) {
      throw new IllegalArgumentException("Settings cannot be null");
    }
    if (analyzer == null) {
      throw new IllegalArgumentException("Analyzer cannot be null");
    }
    this.settings = settings;
    this.analyzer = analyzer;
    this.checkpoint = new Checkpoint(settings.checkpointPath());
  }

  /**
   * Parses the Twitter archive JSON and exports tweets to CSV format.
   * This is the first step in the workflow, transforming raw archive data into a processable
   * format.
   *
   * @throws IOException if an I/O error occurs during parsing or writing
   */
  public void extractTweets() throws IOException {
    LOGGER.info("Reading tweets from " + settings.tweetsArchivePath());
    TweetParser parser = new TweetParser(settings.tweetsArchivePath(), ParserType.JSON);

    List<Tweet> tweets;
    try {
      tweets = parser.parse();
    } catch (IOException e) {
      LOGGER.severe(
          "Failed to parse tweets from "
              + settings.tweetsArchivePath()
              + ": "
              + e.getMessage());
      throw new IOException(
          "Failed to parse tweets from "
              + settings.tweetsArchivePath()
              + " (file may not be valid JSON): "
              + e.getMessage(),
          e);
    }

    LOGGER.info("Extracted " + tweets.size() + " tweets, writing to CSV");

    try (CSVWriter writer = CSVWriter.create(settings.transformedTweetsPath(), false)) {
      writer.writeTweets(tweets);
    } catch (IOException e) {
      LOGGER.severe("Error writing tweets to CSV: " + e.getMessage());
      throw new IOException("Error writing tweets to CSV: " + e.getMessage(), e);
    }

    LOGGER.info(
        "Successfully wrote "
            + tweets.size()
            + " tweets to "
            + settings.transformedTweetsPath());
    System.out.printf("Extracted %d tweets to %s%n", tweets.size(), settings.transformedTweetsPath());
  }

  /**
   * Processes one batch of tweets using the Gemini analyzer.
   * Resumes from the last checkpoint, analyzes batchSize tweets, writes flagged results,
   * and saves progress.
   *
   * @throws IOException if an I/O error occurs during analysis
   */
  public void analyzeTweets() throws IOException {
    LOGGER.info("Loading tweets from " + settings.transformedTweetsPath());
    List<Tweet> tweets;
    try {
      tweets = parseTransformedTweets();
    } catch (IOException e) {
      LOGGER.severe("Failed to parse transformed tweets: " + e.getMessage());
      throw new IOException("Failed to parse transformed tweets: " + e.getMessage(), e);
    }

    if (tweets.isEmpty()) {
      LOGGER.warning("No tweets found to analyze");
      return;
    }

    LOGGER.info("Loaded " + tweets.size() + " tweets for analysis");

    int startIdx;
    try {
      startIdx = checkpoint.load();
    } catch (IOException e) {
      LOGGER.severe("Failed to load checkpoint: " + e.getMessage());
      throw new IOException("Failed to load checkpoint: " + e.getMessage(), e);
    }

    LOGGER.info("Resuming from tweet index " + startIdx);

    if (startIdx >= tweets.size()) {
      LOGGER.info("All tweets already analyzed");
      System.out.println("All tweets already analyzed");
      return;
    }

    int endIdx = Math.min(startIdx + settings.batchSize(), tweets.size());

    LOGGER.info(
        String.format(
            "Processing batch (tweets %d-%d of %d)", startIdx + 1, endIdx, tweets.size()));
    System.out.printf("Processing tweets %d to %d (total: %d)%n", startIdx, endIdx - 1, tweets.size());

    int analyzedCount = 0;
    int deleteCandidates = 0;

    try (CSVWriter writer = CSVWriter.create(settings.processedResultsPath(), true)) {
      for (int i = startIdx; i < endIdx; i++) {
        Tweet tweet = tweets.get(i);

        if (isRetweet(tweet)) {
          continue;
        }

        AnalysisResult result;
        try {
          result = analyzer.analyze(tweet);
          analyzedCount++;
          LOGGER.fine("Tweet " + tweet.id() + ": " + result.decision());
        } catch (Exception e) {
          LOGGER.severe("Failed to analyze tweet " + tweet.id() + ": " + e.getMessage());
          throw new IOException("Failed to analyze tweet " + tweet.id() + ": " + e.getMessage(), e);
        }

        if (result.decision() == Decision.DELETE) {
          deleteCandidates++;
          writer.writeResult(result);
        }
      }
    } catch (IOException e) {
      LOGGER.severe("Failed to write results: " + e.getMessage());
      throw new IOException("Failed to write results: " + e.getMessage(), e);
    }

    try {
      checkpoint.save(endIdx);
      LOGGER.info("Checkpoint saved at index " + endIdx);
    } catch (IOException e) {
      LOGGER.severe("Error saving checkpoint: " + e.getMessage());
      throw new IOException("Error saving checkpoint: " + e.getMessage(), e);
    }

    LOGGER.info(
        String.format(
            "Batch complete! Analyzed %d tweets, found %d deletion candidates (%d/%d total)",
            analyzedCount, deleteCandidates, endIdx, tweets.size()));
    LOGGER.info("Results written to " + settings.processedResultsPath());
    System.out.printf(
        "Batch complete! Processed %d tweets (%d/%d total)%n",
        endIdx - startIdx, endIdx, tweets.size());
  }

  private List<Tweet> parseTransformedTweets() throws IOException {
    TweetParser parser = new TweetParser(settings.transformedTweetsPath(), ParserType.CSV);
    return parser.parse();
  }

  /**
   * Checks if a tweet is a retweet by looking for "RT @" prefix.
   *
   * @param tweet the tweet to check
   * @return true if the tweet is a retweet, false otherwise
   */
  private static boolean isRetweet(Tweet tweet) {
    String content = tweet.content();
    return content != null && content.startsWith("RT @");
  }
}
